### PR TITLE
Refactor project get/set

### LINKED ISSUE #1058: Feature Request: function that tests if renv project is loaded
Use-case:

When opening an .Rproj that uses renv, a message is printed like:
```r
* Project '/path/to/project/' loaded. [renv 0.15.5]
```
I can manually inspect the console output for that message, but it would be nice to add a line to my code that tests that it was successfully loaded. This would be particularly useful for cases where I'm running scripts in batch mode. Then my scripts could include something like:
```r
stopifnot(renv_is_loaded)
```
This came up because I noticed that I was running scripts in batch mode from the wrong directory, so renv was not being sourced properly.

### PR DESCRIPTION
**Pull request by hadley**
* Eliminate `renv_project()` in favour of `renv_project_get()`.

* Use `renv_project_resolve()` in `load()`, like all other functions. Move `renv_project_find()` to `autoload.R`, the only place it's now used.

* Use `renv_project_set()`, `renv_project_get()`, and `renv_project_clear()` everywhere, instead of accessing `RENV_PROJECT` env var directly.

*  Implement `renv_project_is_active()` and use where appropriate. Plan is to export as `is_loaded()` to fix #1058.

* I don't understand how `renv_snapshot_task()` worked because, AFAICT, `project` wasn't previously defined.

### CODE REVIEW

File: R/autoload.R [hunk 1]
```diff
@@ -81,3 +81,24 @@ renv_autoload_impl <- function() {
   TRUE
 
 }
+
+# TODO: this gets really dicey once the user starts configuring where
+# renv places its project-local state ...
+renv_project_find <- function(project = NULL) {
```
Comment (kevinushey) [inline R/autoload.R:87]: Why was this moved to `autoload.R`? By convention, files with the prefix `renv_<name>` and placed in `<name>.R`.

File: R/activate.R [hunk 2]
```diff
@@ -142,7 +142,7 @@ renv_activate_prompt <- function(action, library, prompt, project) {
     prompt &&
     interactive() &&
     is.null(library) &&
-    !identical(project, Sys.getenv("RENV_PROJECT"))
+    !renv_project_is_active(project)
```
Comment (kevinushey) [inline R/activate.R:145]:
```suggestion
    !renv_project_loaded(project)
```

File: R/project.R [hunk 3]
```diff
@@ -18,17 +18,24 @@
...
 
   project
+}
+
+renv_project_resolve <- function(project = NULL) {
+  project <- project %||% renv_project_get()
+  normalizePath(project, winslash = "/", mustWork = FALSE)
+}
 
+renv_project_is_active <- function(project = NULL) {
```
Comment (kevinushey) [inline R/project.R:37]: This doesn't distinguish between "the user appears to have requested a particular project by setting the `RENV_PROJECT` environment variable" and "the project was actually loaded".

File: R/project.R [hunk 4]
```diff
@@ -18,17 +18,24 @@
...
+renv_project_get <- function(default = getwd()) {
   project <- Sys.getenv("RENV_PROJECT", unset = NA)
   if (is.na(project))
     return(default)
 
   project
+}
+
+renv_project_resolve <- function(project = NULL) {
+  project <- project %||% renv_project_get()
```
Comment (kevinushey) [inline R/project.R:33]:
```suggestion
  project <- project %??% renv_project_get()
```

File: R/autoload.R [hunk 5]
```diff
@@ -81,3 +81,24 @@ renv_autoload_impl <- function() {
...
+
+  project <- project %||% getwd()
```
Comment (kevinushey) [inline R/autoload.R:89]:
```suggestion
  project <- project %??% getwd()
```

File: R/load.R [hunk 6]
```diff
@@ -64,8 +60,7 @@ load <- function(project = NULL, quiet = FALSE) {
   # then unload the current project and reload the requested one
   switch <-
     !renv_metadata_embedded() &&
-    !is.na(Sys.getenv("RENV_PROJECT", unset = NA)) &&
-    !identical(project, renv_project())
+    !renv_project_is_active(project)
```
Comment (kevinushey) [inline R/load.R:63]:
```suggestion
    !renv_project_loaded(project)
```

We might also need to do this stuff if any project is loaded (not just the current one)?

File: R/modify.R [hunk 7]
```diff
@@ -38,7 +38,7 @@ renv_modify_impl <- function(project, changes) {
   else
     renv_modify_noninteractive(project, changes)
 
-  if (identical(renv_project(), project))
+  if (renv_project_is_active(project))
```
Comment (kevinushey) [inline R/modify.R:41]:
```suggestion
  if (renv_project_loaded(project))
```

File: R/project.R (hunk 3 above)
Comment (kevinushey) [inline R/project.R:37]:
```suggestion
renv_project_loaded <- function(project = NULL) {
```

File: R/load.R [hunk 8]
```diff
@@ -29,11 +29,7 @@ load <- function(project = NULL, quiet = FALSE) {
 
   renv_scope_error_handler()
 
-  project <- normalizePath(
-    project %||% renv_project_find(project),
-    winslash = "/",
-    mustWork = TRUE
-  )
+  project <- renv_project_resolve(project)
```
Comment (kevinushey) [inline R/load.R:32]: Do we want to preserve the `renv_project_find()` behavior here? (I think we do)

File: R/activate.R (hunk 2 above)
Comment (hadley) [reply R/activate.R:145]: Are you suggesting a different name for this function?

File: R/project.R (hunk 4 above)
Comment (hadley) [reply R/project.R:33]: I'm not sure you want to get into the details, but why have a separate `%??%` and `%||%` and why make `%??%` more like the `%||%` we use everywhere else? 😭

File: R/load.R (hunk 8 above)
Comment (hadley) [reply R/load.R:32]: Yeah, I'll revert.

File: R/autoload.R (hunk 1 above)
Comment (hadley) [reply R/autoload.R:87]: I moved it because that was the only place it was used, but since I'm reverting that change, I'll move it back.

**Comment by hadley** [comment]
> Are we handling this correctly in any of the updated locations?

Yes; `renv_project_get()` is called with an explicit default in just about every location. I'll refactor to make `default` a required argument

> I think we should have a variable `_renv_project_path` that gets set when `renv::load()` is called, and use that wherever we need to know what the currently-loaded project is. (`RENV_PROJECT` is unreliable as it might be set by users, and it could be set even if the project is not yet loaded.)

Why would a user set `RENV_PROJECT`? How would it get set if the project is not yet loaded?

But I do agree that this should be an option, since it doesn't seem like you want subprocesses to inherit this value. If we change to the option, do you see any reason to continue supporting `RENV_PROJECT`?

> There's probably some conversation to be had about the distinction between `RENV_PROJECT` and the `renv.project.path` option.

There's some subtle bug that I haven't quite figured out due to `RENV_PROJECT` being set at the start of project loading and `renv.project.path` being set at the end. Assuming I can resolve this do you see any issue with eliminating that option?

File: R/activate.R (hunk 2 above)
Comment (kevinushey) [reply R/activate.R:145]: Yeah, that's my intent here.

**Comment by kevinushey** [comment]
> Why would a user set RENV_PROJECT? How would it get set if the project is not yet loaded?

Good question. I guess I was imagining something like:

```
RENV_PROJECT=/path/to/project R -e 'renv::run(...)'
```

That is, having some way to load a project in a different location, but run using some separate working directory in an ergonomic way.

Now that I say it out loud, I'm not sure if anyone is actually using it this way...

> But I do agree that this should be an option, since it doesn't seem like you want subprocesses to inherit this value. If we change to the option, do you see any reason to continue supporting RENV_PROJECT?

I think that makes sense ... although I wonder what we should do in this scenario:

1. You work in an `renv` project, that project is loaded,
2. You `setwd("scripts")` or otherwise change the working directory,
3. You run something like `callr::r(function() { ... })`

Should that sub-process run with the same project loaded by default?

**Comment by kevinushey** [comment]
> There's some subtle bug that I haven't quite figured out due to RENV_PROJECT being set at the start of project loading and renv.project.path being set at the end. Assuming I can resolve this do you see any issue with eliminating that option?

I think that makes sense. The intention was for `renv.project.path` to be set at the end of `renv::load()` so we could distinguish between the "requested" project and the "actually currently loaded" project.

But then I sort of feel like if this is something handled internally by `renv`, it should just be a variable in the `renv` namespace rather than an option.

**Comment by hadley** [comment]
FWIW, as far as I can tell, `RENV_PROJECT` isn't documented anywhere, so it should be ok to remove.

In the callr scenario, it defaults to copying the `.libPaths()` from the current session, so you'll at least load packages from the right place. But probably `RENV_PROJECT` shouldn't be set, since we haven't actually loaded renv there.

**Comment by kevinushey** [comment]
Sorry, I just remembered we're using it in the IDE:

https://github.com/rstudio/rstudio/blob/0e442b55fae1003f82bc3dc1cde2f037dc8815db/src/cpp/session/modules/SessionRenv.cpp#L39-L42

so we need to at least keep it alive for some amount of time for IDE compatibility.

**Comment by kevinushey** [comment]
> I don't understand how renv_snapshot_task() worked because, AFAICT, project wasn't previously defined.

That looks like a bug; I'll see if I can fix.
